File: modules/teleprompter-pip/android/src/main/java/expo/modules/teleprompterpip/TeleprompterPipModule.kt
package expo.modules.teleprompterpip

import android.content.Intent
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class TeleprompterPipModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TeleprompterPip")

    AsyncFunction("startTeleprompterPip") { config: Map<String, Any?> ->
      val context = appContext.reactContext ?: throw IllegalStateException("React context unavailable")
      val text = (config["text"] as? String).orEmpty()
      val fontSize = (config["fontSize"] as? Number)?.toFloat() ?: 24f
      val scrollSpeed = (config["scrollSpeed"] as? Number)?.toFloat() ?: 40f
      val prepDelay = (config["preparationDelaySeconds"] as? Number)?.toInt() ?: 3

      val intent = TeleprompterPipActivity.createIntent(
        context,
        text,
        fontSize,
        scrollSpeed,
        prepDelay
      )

      intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      context.startActivity(intent)
    }

    AsyncFunction("stopTeleprompterPip") {
      TeleprompterPipActivity.finishCurrent()
    }
  }
}

File: modules/teleprompter-pip/src/TeleprompterPipModule.ts
import { requireOptionalNativeModule } from 'expo-modules-core';

export interface TeleprompterPipConfig {
  text: string;
  fontSize: number;
  fontFamily?: string;
  scrollSpeed: number;
  preparationDelaySeconds: number;
  width?: number;
  height?: number;
  autoBackgroundAfterStart?: boolean;
}

interface TeleprompterPipModuleType {
  startTeleprompterPip(config: TeleprompterPipConfig): Promise<void>;
  stopTeleprompterPip(): Promise<void>;
}

const nativeModule = requireOptionalNativeModule<TeleprompterPipModuleType>('TeleprompterPip');

const notAvailable = (): Promise<never> =>
  Promise.reject(new Error('TeleprompterPip native module is not available on this platform'));

const TeleprompterPipModule: TeleprompterPipModuleType = {
  startTeleprompterPip: nativeModule
    ? (config) => nativeModule.startTeleprompterPip(config)
    : () => notAvailable(),
  stopTeleprompterPip: nativeModule
    ? () => nativeModule.stopTeleprompterPip()
    : () => notAvailable(),
};

export default TeleprompterPipModule;
